File: MongoDBSession/Endpoints/MongoDb.cs
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Driver;
using MongoDBSession.Models;

namespace MongoDBSession.Endpoints;

public class MongoDbSession
{
    [BsonId]
    public string SessionId { get; set; }
    public string UserId { get; set; }
    public string Name { get; set; }
    public string Email { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime ExpiresAt { get; set; }
}

public static class MongoDbLogin
{
    public static void MapMongoDbLogin(this IEndpointRouteBuilder app)
    {
        app.MapPost("/api/mongodb/login", async (IMongoDatabase database) =>
        {
            var randomUser = await MemberService.GetRandomMemberAsync(database);

            var sessionKey = Guid.NewGuid().ToString();

            var sessionsCollection = database.GetCollection<MongoDbSession>("Sessions");

            var keys = Builders<MongoDbSession>.IndexKeys.Ascending(doc => doc.ExpiresAt);
            var options = new CreateIndexOptions { ExpireAfter = TimeSpan.Zero };
            var indexModel = new CreateIndexModel<MongoDbSession>(keys, options);
            await sessionsCollection.Indexes.CreateOneAsync(indexModel);

            var doc = new MongoDbSession
            {
                SessionId = sessionKey,
                UserId = randomUser.Id,
                Name = randomUser.Name,
                Email = randomUser.Email,
                CreatedAt = DateTime.UtcNow,
                ExpiresAt = DateTime.UtcNow.AddMinutes(Const.ExpiryMinutes)
            };

            await sessionsCollection.ReplaceOneAsync(
                filter: d => d.SessionId == sessionKey,
                replacement: doc,
                options: new ReplaceOptions { IsUpsert = true }
            );

            return Results.Ok(new
            {
                Token = sessionKey,
                ExpiresInMinutes = Const.ExpiryMinutes,
                User = new
                {
                    Id = randomUser.Id,
                    Name = randomUser.Name,
                    Email = randomUser.Email
                }
            });
        });
    }
}



public static class MongoDbProfile
{
    public static void MapMongoDbProfile(this IEndpointRouteBuilder app)
    {
        app.MapGet("/api/mongodb/profile", async (HttpContext httpContext, IMongoDatabase database) =>
        {
            var authHeader = httpContext.Request.Headers["Authorization"].ToString();
            var token = authHeader.Substring("Bearer ".Length).Trim();
            var collection = database.GetCollection<MongoDbSession>("Sessions");

            var doc = await collection.Find(d => d.SessionId == token).FirstOrDefaultAsync();
            if (doc == null)
                return Results.Json(new { Message = "找不到對應的 MongoDB Session，或該 Session 已經過期！" }, statusCode: StatusCodes.Status401Unauthorized);

            return Results.Ok(new
            {
                Message = "恭喜！您已成功通過 MongoDB 驗證，成功存取受保護的 Profile 路由。",
                User = new
                {
                    Id = doc.UserId,
                    Name = doc.Name,
                    Email = doc.Email
                },
            });
        });
    }
}

File: MongoDBSession/src/endpoints/mongodb.js
const express = require('express');
const crypto = require('crypto');
const MongoDbSession = require('../models/MongoDbSession');
const MemberService = require('../services/MemberService');
const Const = require('../constants');

const router = express.Router();

router.post('/login', async (req, res) => {
    const randomUser = await MemberService.getRandomMemberAsync();

    const sessionKey = crypto.randomUUID();
    const sessionDoc = {
        SessionId: sessionKey,
        UserId: randomUser._id,
        Name: randomUser.Name,
        Email: randomUser.Email,
        CreatedAt: new Date(),
        ExpiresAt: new Date(Date.now() + Const.ExpiryMinutes * 60000)
    };

    await MongoDbSession.findOneAndReplace(
        { SessionId: sessionKey },
        sessionDoc,
        { upsert: true }
    );

    res.json({
        Token: sessionKey,
        ExpiresInMinutes: Const.ExpiryMinutes,
        User: {
            Id: randomUser._id,
            Name: randomUser.Name,
            Email: randomUser.Email
        }
    });
});

router.get('/profile', async (req, res) => {
    const authHeader = req.headers.authorization;

    const token = authHeader.split(' ')[1];
    const session = await MongoDbSession.findOne({ SessionId: token });

    if (!session || new Date() > session.ExpiresAt) {
        return res.status(401).json({ Message: "找不到對應的 MongoDB Session，或該 Session 已經過期！" });
    }

    res.json({
        Message: "恭喜！您已成功通過 MongoDB 驗證，成功存取受保護的 Profile 路由。",
        User: {
            Id: session.UserId,
            Name: session.Name,
            Email: session.Email
        }
    });
});

module.exports = router;
